### PR TITLE
Add support for caching tablegen

### PR DESCRIPTION
**Pull request by dexonsmith**
In git-log order:
- 63a48d5d2b7b TableGen: Add caching using the CAS
- 8076dd5f1e75 TableGen: Split out a TableGenImpl that does not need Argv0, NFC
- fe64b850352e TableGen: Start passing all args to TableGenMain, NFC
- 20f2319f42c5 Support: Add PrefixMapper and RealPathPrefixMapper
- dedbfc95b56c WIP: Support: Expose sys::path::system_style()
- ac9fd93d489b Support: Add Expected::moveInto and Expected::emplaceInto

Uses the `SourceMgr` change from https://github.com/apple/llvm-project/pull/3402.

### CODE REVIEW

File: llvm/include/llvm/Support/Error.h [hunk 1]
```diff
@@ -577,6 +577,30 @@ template <class T> class LLVM_NODISCARD Expected {
     return const_cast<Expected<T> *>(this)->get();
   }
 
+  /// Returns \a takeError() after moving the held T (if any) into \p V.
```
Comment (cachemeifyoucan) [inline llvm/include/llvm/Support/Error.h:580]: Maybe this can go directly into upstream?

File: llvm/lib/TableGen/Main.cpp [hunk 2]
```diff
@@ -111,25 +150,52 @@ class AlreadyReportedError : public ErrorInfo<AlreadyReportedError> {
...
+    // This prevents recompilation of all the files depending on it if there
+    // aren't any.
+    if (auto ExistingOrErr =
+            MemoryBuffer::getFile(OutputFilename, /*IsText=*/true))
+      if (std::move(ExistingOrErr.get())->getBuffer() == Out)
+        WriteFile = false;
+  }
+  if (WriteFile) {
+    std::error_code EC;
+    ToolOutputFile OutFile(OutputFilename, EC, sys::fs::OF_Text);
```
Comment (cachemeifyoucan) [inline llvm/lib/TableGen/Main.cpp:166]: No `OutputBackend` here? 😄

**Comment by cachemeifyoucan** [comment]
LGTM. First commit can probably go straight into upstream. Some other NFC changes will be good to go upstream as well.

TableGen changes LGTM.

**Review by akyrtzi — APPROVED** [review]
I think the tablegen dependencies scanning functionality would be goodness to go upstream on its own, since it can benefit Bazel-like build systems that want to have the dependencies in advance.

**Comment by dexonsmith** [comment]
> LGTM. First commit can probably go straight into upstream.

Yup, I was preparing for that; posted at: https://reviews.llvm.org/D112278. Rebased this one to remove extraneous changes.

> Some other NFC changes will be good to go upstream as well.

Good point, I'll work on it.

> TableGen changes LGTM.

Thanks for the review!

**Review by cachemeifyoucan — APPROVED** [review]

**Comment by dexonsmith** [comment]
> I think the tablegen dependencies scanning functionality would be goodness to go upstream on its own, since it can benefit Bazel-like build systems that want to have the dependencies in advance.

Agreed that it would be... might need to separate out another function or two from the CAS logic / caching / etc but probably not hard... The next step would be building an actual tool for Bazel to call though; not sure I have time to figure that out right now but we can come back to it. I'll add a FIXME.

File: llvm/lib/TableGen/Main.cpp (hunk 2 above)
Comment (dexonsmith) [reply llvm/lib/TableGen/Main.cpp:166]: It was easier to just refactor the existing TableGen code as a one-off... not ideal!

**Comment by akyrtzi** [comment]
> The next step would be building an actual tool for Bazel to call though;

Can it be just tablegen itself? I mean invoking tablegen with an extra flag that would make it dump or write out the scanned dependencies for that invocation.

**Comment by dexonsmith** [comment]
> > The next step would be building an actual tool for Bazel to call though;
> 
> Can it be just tablegen itself? I mean invoking tablegen with an extra flag that would make it dump or write out the scanned dependencies for that invocation.

Yup, probably could be `*-tblgen` themselves. Caveat is that it's a bit more dangerous that logic can get out of sync if the subsequent invocation doesn't use the CAS tree that was scanned. When it only gets to see the CAS tree, "missed" dependencies don't even make it to the computation, so the build system doesn't have to worry about correctness.

**Comment by dexonsmith** [comment]
BTW, fixed the commit message here to actually give details: https://github.com/apple/llvm-project/commit/2155273204bc3a49d8cf406be9dd67694fc98aa8

**Comment by akyrtzi** [comment]
> Caveat is that it's a bit more dangerous that logic can get out of sync if the subsequent invocation doesn't use the CAS tree that was scanned. When it only gets to see the CAS tree, "missed" dependencies don't even make it to the computation, so the build system doesn't have to worry about correctness.

Technically the build system could verify that only the expected files were used by comparing with the `.d` file output and/or running tablegen in a sandbox.

**Pull request merged**